ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -579,8 +583,18 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
       boolean tokenWasReset = false;
       Offset logEndOffsetBeforeFind = log.getEndOffset();
       StoreFindToken storeToken = (StoreFindToken) token;
+      UUID remoteIncarnationId = storeToken.getIncarnationId();
+      if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized) &&
+          remoteIncarnationId != null && remoteIncarnationId.compareTo(incarnationId) != 0) {
+        // incarnationId mismatch, hence resetting the token to beginning
+        logger.info("Index : {} resetting offset after incarnation, new incarnation Id {}, "
+            + "incarnationId from store token {}", dataDir, incarnationId, remoteIncarnationId);
+        storeToken = new StoreFindToken(incarnationId);
+        tokenWasReset = true;
+      }
+      // if incarnationId is null, for backwards compatability purposes, will consider the token as good
```

Comment:
I think this "pre-check" has become sufficiently big to warrant a method of its own. What do you think? It could do these checks and return true/false based on whether token was reset.